File: backend/app/routers/game.py
from datetime import date, timedelta

from fastapi import APIRouter, Depends, Header, HTTPException
from sqlalchemy.orm import Session, joinedload

from app.core.time import get_client_today
from app.db.session import get_db
from app.models.card import Card
from app.models.daily_answer import DailyAnswer
from app.models.guess import Guess
from app.schemas.guess import (
    GuessRequest,
    GuessResponse,
    PastGuess,
    PreviousAnswerResponse,
    TodayGuessesResponse,
    TodayWinnersResponse,
)
from app.services.daily_answer import get_or_create_daily_answer
from app.services.game import MAX_GUESSES, compare_cards, is_correct_guess

router = APIRouter(prefix="/game", tags=["game"])


@router.post("/guess", response_model=GuessResponse)
def guess(
    payload: GuessRequest,
    db: Session = Depends(get_db),
    today: date = Depends(get_client_today),
    # Identifies this browser without requiring login — generated and stored
    # by the client itself (see frontend/src/utils/guestSession.ts), not a
    # cookie the backend sets. A cross-site cookie here (frontend and backend
    # are on different domains) gets silently blocked by Safari's Intelligent
    # Tracking Prevention regardless of SameSite=None; Secure, which broke
    # guess-restore-on-refresh for some iOS players; a plain header isn't
    # subject to that at all.
    guest_session_id: str = Header(alias="X-Guest-Session-Id"),
):
    guessed_card = db.query(Card).filter(Card.name == payload.guess_name).first()
    if guessed_card is None:
        raise HTTPException(status_code=404, detail=f"No card named '{payload.guess_name}'")

    daily_answer = get_or_create_daily_answer(db, today)
    secret_card = daily_answer.card

    past_guesses = (
        db.query(Guess)
        .filter(Guess.daily_answer_id == daily_answer.id, Guess.guest_session_id == guest_session_id)
        .all()
    )
    if any(g.is_correct for g in past_guesses):
        raise HTTPException(status_code=400, detail="Already guessed today's card correctly")
    if len(past_guesses) >= MAX_GUESSES:
        raise HTTPException(status_code=400, detail="Out of guesses for today")

    comparisons = compare_cards(secret_card, guessed_card)
    correct = is_correct_guess(comparisons)

    db.add(
        Guess(
            daily_answer_id=daily_answer.id,
            guessed_card_id=guessed_card.id,
            guest_session_id=guest_session_id,
            is_correct=correct,
        )
    )
    db.commit()

    # This guess is the one that used up the last try without winning —
    # reveal the answer now instead of waiting for a refresh.
    out_of_guesses = not correct and len(past_guesses) + 1 >= MAX_GUESSES
    reveal_answer = secret_card.name if out_of_guesses else None

    return GuessResponse(comparisons=comparisons, is_correct=correct, reveal_answer=reveal_answer)


@router.get("/today", response_model=TodayGuessesResponse)
def today_guesses(
    db: Session = Depends(get_db),
    today: date = Depends(get_client_today),
    guest_session_id: str | None = Header(default=None, alias="X-Guest-Session-Id"),
):
    """Replays this browser's guesses for today's answer, so a page refresh
    doesn't lose progress. Comparisons aren't stored on the Guess row — they're
    just recomputed here the same way /guess computed them originally, since
    they're a pure function of (secret card, guessed card)."""
    if not guest_session_id:
        # Never guessed on this browser before (or an old cached client that
        # predates the guest-session header — see guess() above) — nothing
        # to restore, and nothing worth touching the DB for yet.
        return TodayGuessesResponse(guesses=[])

    daily_answer = get_or_create_daily_answer(db, today)
    secret_card = daily_answer.card

    past_guesses = (
        db.query(Guess)
        .options(joinedload(Guess.guessed_card))
        .filter(
            Guess.daily_answer_id == daily_answer.id,
            Guess.guest_session_id == guest_session_id,
        )
        .order_by(Guess.created_at.asc())
        .all()
    )

    lost = len(past_guesses) >= MAX_GUESSES and not any(g.is_correct for g in past_guesses)

    return TodayGuessesResponse(
        guesses=[
            PastGuess(
                card_name=g.guessed_card.name,
                comparisons=compare_cards(secret_card, g.guessed_card),
                is_correct=g.is_correct,
            )
            for g in past_guesses
        ],
        reveal_answer=secret_card.name if lost else None,
    )


@router.get("/today/winners", response_model=TodayWinnersResponse)
def today_winners(db: Session = Depends(get_db), today: date = Depends(get_client_today)):
    """How many distinct guest sessions have correctly guessed today's secret
    card so far — public, not tied to this browser's own guest session. Counts
    distinct guest_session_id rather than distinct Guess rows so someone who
    somehow submits the correct guess more than once still only counts once.
    "Today" is this requesting client's own timezone (see core/time.py), so
    this is specifically "winners of the card you're playing", not a single
    worldwide count — players in other timezones may be on a different card
    entirely right now."""
    daily_answer = get_or_create_daily_answer(db, today)

    winners_count = (
        db.query(Guess.guest_session_id)
        .filter(Guess.daily_answer_id == daily_answer.id, Guess.is_correct.is_(True))
        .distinct()
        .count()
    )

    return TodayWinnersResponse(winners_count=winners_count)


@router.get("/previous-answer", response_model=PreviousAnswerResponse)
def previous_answer(db: Session = Depends(get_db), today: date = Depends(get_client_today)):
    """Yesterday's secret card, for the small footer line. Looked up directly
    rather than via get_or_create_daily_answer — a missing row here just means
    nothing to show, not something to generate (unlike today's answer, which
    the game needs to exist)."""
    yesterday = today - timedelta(days=1)
    daily_answer = (
        db.query(DailyAnswer)
        .options(joinedload(DailyAnswer.card))
        .filter(DailyAnswer.date == yesterday)
        .first()
    )
    return PreviousAnswerResponse(card_name=daily_answer.card.name if daily_answer else None)

File: frontend/src/api/game.ts
// Mirrors backend/app/schemas/guess.py — keep these in sync if that changes.

import { getGuestSessionId } from '../utils/guestSession'

export type StatComparison = 'match' | 'mismatch' | 'higher' | 'lower'

export interface GuessResult {
  comparisons: Record<string, StatComparison>
  is_correct: boolean
  // Set only on the guess that uses up the last of 8 tries without winning.
  reveal_answer: string | null
}

export interface PastGuess {
  card_name: string
  comparisons: Record<string, StatComparison>
  is_correct: boolean
}

export interface TodayGuesses {
  guesses: PastGuess[]
  // Set when this session already lost today's game before this page load.
  reveal_answer: string | null
}

export interface PreviousAnswer {
  card_name: string | null
}

export interface TodayWinners {
  winners_count: number
}

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL ?? 'http://localhost:8000'

// Which day's card this browser plays is decided by this timezone, not a
// fixed one on the backend — see core/time.py's get_client_today. Read once;
// a player's timezone doesn't change mid-session.
const TIMEZONE_HEADERS = { 'X-Timezone': Intl.DateTimeFormat().resolvedOptions().timeZone }

// Identifies this browser's guesses to the backend — see guestSession.ts for
// why this is a header the client attaches itself rather than a cookie the
// backend sets. Read once; stable for the life of the page the same way
// TIMEZONE_HEADERS is.
const GUEST_SESSION_HEADERS = { 'X-Guest-Session-Id': getGuestSessionId() }

export async function submitGuess(cardName: string): Promise<GuessResult> {
  const response = await fetch(`${API_BASE_URL}/game/guess`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json', ...TIMEZONE_HEADERS, ...GUEST_SESSION_HEADERS },
    body: JSON.stringify({ guess_name: cardName }),
  })

  if (!response.ok) {
    const body = await response.json().catch(() => null)
    throw new Error(body?.detail ?? `Guess request failed (${response.status})`)
  }

  return response.json()
}

export async function fetchTodayGuesses(): Promise<TodayGuesses> {
  const response = await fetch(`${API_BASE_URL}/game/today`, {
    headers: { ...TIMEZONE_HEADERS, ...GUEST_SESSION_HEADERS },
  })

  if (!response.ok) {
    throw new Error(`Failed to load today's guesses (${response.status})`)
  }

  return response.json()
}

export async function fetchPreviousAnswer(): Promise<PreviousAnswer> {
  const response = await fetch(`${API_BASE_URL}/game/previous-answer`, {
    headers: TIMEZONE_HEADERS,
  })

  if (!response.ok) {
    throw new Error(`Failed to load previous answer (${response.status})`)
  }

  return response.json()
}

export async function fetchTodayWinners(): Promise<TodayWinners> {
  // Public count, not tied to this browser's guest session — no cookie needed.
  // Still needs the timezone header though, since "today" now depends on it.
  const response = await fetch(`${API_BASE_URL}/game/today/winners`, {
    headers: TIMEZONE_HEADERS,
  })

  if (!response.ok) {
    throw new Error(`Failed to load today's winners count (${response.status})`)
  }

  return response.json()
}
